File: articles/desalination-plant-security/julia/desalination_resilience_analysis.jl
# Desalination resilience analysis in Julia.
# Synthetic demonstration only.

using CSV
using DataFrames

root = normpath(joinpath(@__DIR__, ".."))
data_path = joinpath(root, "data", "raw", "desalination_systems_synthetic.csv")
out_path = joinpath(root, "outputs", "tables", "desalination_resilience_summary_julia.csv")

df = CSV.read(data_path, DataFrame)

df.normal_available_supply_mld =
    df.normal_desal_output_mld .+
    df.alternative_supply_mld .+
    df.emergency_transfer_mld .-
    df.system_losses_mld

df.outage_desal_output_mld =
    (1 .- df.outage_fraction) .* df.normal_desal_output_mld

df.outage_available_supply_mld =
    df.outage_desal_output_mld .+
    df.alternative_supply_mld .+
    df.emergency_transfer_mld .-
    df.system_losses_mld

df.desalination_dependency_ratio =
    df.normal_desal_output_mld ./ (
        df.normal_desal_output_mld .+
        df.alternative_supply_mld .+
        df.emergency_transfer_mld
    )

df.meets_priority_demand_under_outage =
    df.outage_available_supply_mld .>= df.priority_demand_mld

df.daily_deficit_mld =
    max.(df.priority_demand_mld .- df.outage_available_supply_mld, 0)

df.storage_coverage_days =
    ifelse.(df.daily_deficit_mld .> 0, df.storage_reserve_mld ./ df.daily_deficit_mld, Inf)

df.storage_covers_recovery_period =
    (df.daily_deficit_mld .== 0) .| (df.storage_coverage_days .>= df.recovery_days)

mkpath(dirname(out_path))
CSV.write(out_path, df)

println(df[:, [
    :city,
    :plant,
    :priority_demand_mld,
    :outage_available_supply_mld,
    :desalination_dependency_ratio,
    :daily_deficit_mld,
    :storage_coverage_days,
    :storage_covers_recovery_period
]])

File: articles/desalination-plant-security/r/desalination_resilience_analysis.R
# Desalination resilience analysis with synthetic data.
# Computes dependency ratios, outage supply, deficits, and storage coverage.

library(readr)
library(dplyr)

root <- normalizePath(file.path(dirname(sys.frame(1)$ofile), ".."))
data_path <- file.path(root, "data", "raw", "desalination_systems_synthetic.csv")
out_path <- file.path(root, "outputs", "tables", "desalination_resilience_summary_r.csv")
processed_path <- file.path(root, "data", "processed", "desalination_resilience_classified_r.csv")

df <- read_csv(data_path, show_col_types = FALSE) %>%
  mutate(
    normal_available_supply_mld =
      normal_desal_output_mld +
      alternative_supply_mld +
      emergency_transfer_mld -
      system_losses_mld,
    outage_desal_output_mld =
      (1 - outage_fraction) * normal_desal_output_mld,
    backup_power_adjusted_output_mld =
      normal_desal_output_mld * backup_power_fraction,
    outage_available_supply_mld =
      outage_desal_output_mld +
      alternative_supply_mld +
      emergency_transfer_mld -
      system_losses_mld,
    desalination_dependency_ratio =
      normal_desal_output_mld /
      (normal_desal_output_mld + alternative_supply_mld + emergency_transfer_mld),
    meets_priority_demand_under_outage =
      outage_available_supply_mld >= priority_demand_mld,
    daily_deficit_mld =
      pmax(priority_demand_mld - outage_available_supply_mld, 0),
    storage_coverage_days =
      if_else(daily_deficit_mld > 0, storage_reserve_mld / daily_deficit_mld, Inf),
    storage_covers_recovery_period =
      daily_deficit_mld == 0 | storage_coverage_days >= recovery_days
  )

summary <- df %>%
  select(
    city,
    plant,
    priority_demand_mld,
    normal_available_supply_mld,
    outage_available_supply_mld,
    desalination_dependency_ratio,
    daily_deficit_mld,
    storage_coverage_days,
    recovery_days,
    storage_covers_recovery_period,
    meets_priority_demand_under_outage
  )

dir.create(dirname(out_path), recursive = TRUE, showWarnings = FALSE)
dir.create(dirname(processed_path), recursive = TRUE, showWarnings = FALSE)

write_csv(df, processed_path)
write_csv(summary, out_path)

print(summary)
